### DISCUSSION TITLE
hook read方法调用BYTEHOOK_POP_STACK()出现崩溃

### DISCUSSION THREAD

**Discussion by dimly123** [bytedance/bhook#38] [Q&A]
static bytehook_stub_t read_stub = NULL;
typedef ssize_t (*read_t)(int, void *, size_t);
static int proxy_read(int fd, void *buf, size_t count) {
    BYTEHOOK_CALL_PREV(proxy_read, read_t, fd, buf, count);
//    __android_log_print(ANDROID_LOG_DEBUG, "jzx-->", "%zd", count);
//    __android_log_print(ANDROID_LOG_DEBUG, "jzx-->", "%p", buf);
//    (void)buf;
//    (void)count;
    BYTEHOOK_POP_STACK();
    return fd;
}
static int hacker_hook(JNIEnv *env, jobject thiz) {
    (void) env, (void) thiz;
    void *read_proxy;
    read_proxy = (void *) proxy_read;
    if (NULL != read_stub) return -1;
    read_stub = bytehook_hook_partial(allow_filter,
                                      NULL, NULL, "read", read_proxy,
                                      NULL, NULL);
    return 0;
}
static int hacker_unhook(JNIEnv *env, jobject thiz) {
    (void) env, (void) thiz;

    if(NULL !=read_stub){
        bytehook_unhook(read_stub);
        read_stub = NULL;
    }
    return 0;
}

**Comment by dimly123** [comment]
我在使用项目的过程中碰到如标题的崩溃，以下是相关信息

问题描述:
我在自己工程引入bhook， 想hook所有的文件read函数实现对文件解密操作。
我在bhook的sample工程里面hook open函数是可以的，但是hook read函数出现崩溃问题，但是有部分日志打印。
文档里介绍c语言要在代理函数return之前调用BYTEHOOK_POP_STACK，在hook read中取消掉BYTEHOOK_POP_STACK未崩溃

**Comment by zhbzhbzhbz** [comment]
@dimly123 哈哈我们俩都犯了低级错误，因为返回值错了。正确代码如下：
```
static ssize_t proxy_read(int fd, void *buf, size_t count) {
    ssize_t bytes_read = BYTEHOOK_CALL_PREV(proxy_read, read_t, fd, buf, count);
    BYTEHOOK_POP_STACK();
    return bytes_read; //此处不应返回fd
}
```
read定义详见：https://man7.org/linux/man-pages/man2/read.2.html

**Reply by xmutzlq** [reply]
I/bytehook_tag: >>>>> hooked. stub: 79f0a76780, status: 13, caller_path_name: /system/lib64/libandroid.so, sym_name: read, new_func: 79f05338fc, prev_func: 0, arg: 0
为什么我用read，status老是返回13呢？

**Reply by zhbzhbzhbz** [reply]
> I/bytehook_tag: >>>>> hooked. stub: 79f0a76780, status: 13, caller_path_name: /system/lib64/libandroid.so, sym_name: read, new_func: 79f05338fc, prev_func: 0, arg: 0 为什么我用read，status老是返回13呢？

> 针对 hook 单个调用者库（single）的情况。如果该库中没有找到需要 hook 的目标符号，会返回这个错误码。（hook “全部”或“部分”调用者库时，如果发现某个库中没有找到目标符号，则不会返回错误）
